### PR TITLE
BUG: assignment to multiple columns when some column do not exist

### LINKED ISSUE #13658: Assignment to multiple columns only works if they existed before
```
import pandas as pd
df = pd.DataFrame({'a': [1, 2]})
df['b'] = 3 # creates column 'b'
df[['a', 'b']] = 4 # assigns to columns 'a' and 'b'
df[['c', 'd']] = 5 # KeyError: "['c' 'd'] not in index" !
df[['b', 'c']] = 6 # KeyError: "['c'] not in index" !
```

I would have expected all the above cases to work, but for some reason the last two fail.  New column creation only seems to work for a single column, whilst multiple-column assignment works only if all columns exist.

> INSTALLED VERSIONS
> python: 3.5.1.final.0
> python-bits: 64
> pandas: 0.18.1
> numpy: 1.11.1

### PR DESCRIPTION
**Pull request by howsiwei**
- [x] closes #13658
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

Previous PR: #26534

In particular, the following code now behaves correctly.

```python
import pandas as pd
df = pd.DataFrame({'a': [0, 1, 2], 'b': [3, 4, 5]})
df[['a', 'c']] = 1
print(df)
```
v0.22: error

master: column `'c'` is converted to index `-1`, which causes last column to be overwritten.
```
   a  b
0  1  1
1  1  1
2  1  1
```

After this PR:
```
   a  b  c
0  1  3  1
1  1  4  1
2  1  5  1
```

### CODE REVIEW

File: pandas/core/indexing.py [hunk 1]
```diff
@@ -201,6 +202,21 @@ def _get_setitem_indexer(self, key):
     def __setitem__(self, key, value):
         if isinstance(key, tuple):
             key = tuple(com.apply_if_callable(x, self.obj) for x in key)
+            # if `key` is an index to multiple columns by name, add each column
+            # if it does not already exist
```
Comment (jreback) [inline pandas/core/indexing.py]: this might be better to do in _get_setitem_indexer

File: pandas/core/frame.py [hunk 2]
```diff
@@ -3007,6 +3007,12 @@ def _setitem_array(self, key, value):
                 for k1, k2 in zip(key, value.columns):
                     self[k1] = value[k2]
             else:
+                if all(is_hashable(k) for k in key):
```
Comment (jreback) [inline pandas/core/frame.py]: so can you add this in a routine on self.loc

call it 
_ensure_listlike_indexers(key, axis=1)

**Review by jreback — CHANGES_REQUESTED** [review]
tests look good and conceptually ideas are find, but need to add the code in reasonable places w/o dealing with complexity of core/indexing.py

some suggestions in-line

File: pandas/core/frame.py (hunk 2 above)
Comment (howsiwei) [reply pandas/core/frame.py]: Currently this code always assume that `axis` = 1. Do we need to allow adding multiple rows via assignment too?

File: pandas/core/indexing.py (hunk 1 above)
Comment (howsiwei) [reply pandas/core/indexing.py]: done

File: pandas/core/indexing.py [hunk 3]
```diff
@@ -165,7 +166,28 @@ def _get_loc(self, key: int, axis: int):
     def _slice(self, obj, axis: int, kind=None):
         return self.obj._slice(obj, axis=axis, kind=kind)
 
+    def _ensure_listlike_indexer(self, key, axis):
```
Comment (jreback) [inline pandas/core/indexing.py]: pls type these arguments & add a doc-string. make sure to indicate that this is a mutating operation.

File: pandas/core/indexing.py [hunk 4]
```diff
@@ -165,7 +166,28 @@ def _get_loc(self, key: int, axis: int):
...
+                try:
+                    self.obj[k]
+                except KeyError:
+                    self.obj[k] = np.nan
+
     def _get_setitem_indexer(self, key):
+        if (
+            self.name == "loc"  # column is indexed by name
+            and isinstance(key, tuple)
+            and len(key) >= 2  # key is at least 2-dimensional
```
Comment (jreback) [inline pandas/core/indexing.py]: you have a lot of condictions here, when are you trying to ensure this is being called?

File: pandas/tests/frame/indexing/test_indexing.py [hunk 5]
```diff
@@ -202,6 +202,29 @@ def test_setitem_list_of_tuples(self, float_frame):
...
+            lambda x: [1, 2],
+            lambda x: np.array([1, 2]),
+            lambda x: x[["B", "C"]],
+            lambda x: x[["B", "A"]].values,
+            lambda x: x[["A", "C"]].values.tolist(),
+        ],
+    )
+    def test_setitem_list_missing_columns(self, float_frame, columns, box):
+        # GH 26534
+        result = float_frame.copy()
```
Comment (jreback) [inline pandas/tests/frame/indexing/test_indexing.py]: I would rather you not construct the expected frame like this, rather include it as another argument in the parameterization; essentially hard code it.

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/indexing.py (hunk 4 above)
Comment (howsiwei) [reply pandas/core/indexing.py]: `self.name == "loc"` ensures that this indexer is is indexed by name (yes for `loc`, no for `ix`).
`isinstance(key, tuple) and len(key) >= 2` filter out indexers with only 1 dimension.
`is_list_like_indexer(key[1])` ensures that the indexer is indexed by multiple columns.
`not com.is_bool_indexer(key[1])` ensures that the indexer is not a boolean indexer.

File: pandas/core/indexing.py [hunk 6]
```diff
@@ -165,7 +166,37 @@ def _get_loc(self, key: int, axis: int):
     def _slice(self, obj, axis: int, kind=None):
         return self.obj._slice(obj, axis=axis, kind=kind)
 
+    def _ensure_listlike_indexer(self, key, axis: int):
```
Comment (jreback) [inline pandas/core/indexing.py]: type key (ok for a followon)

File: pandas/core/indexing.py (hunk 4 above)
Comment (jreback) [reply pandas/core/indexing.py]: then do this in the _LocIndexer (and call super); we don't want to dispatch on self.name like this (and .ix is going away shortly anyways).

Comment (jreback) [reply pandas/core/indexing.py]: bigger questions, is why is this needed if you are *already* doing this for frames above.

Exactly which test hits this?

File: pandas/tests/indexing/test_loc.py [hunk 7]
```diff
@@ -750,6 +750,64 @@ def test_loc_setitem_with_scalar_index(self, indexer, value):
 
         assert is_scalar(result) and result == "Z"
 
+    @pytest.mark.parametrize(
```
Comment (jreback) [inline pandas/tests/indexing/test_loc.py]: are these duplicating the frame tests above?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/indexing/test_loc.py (hunk 7 above)
Comment (howsiwei) [reply pandas/tests/indexing/test_loc.py]: As far as I know frame only allows setting the whole column whereas using indexer allows setting the column only for certain rows.
frame:
```
df[["A", "B"]] = 1
```
indexer:
```
df.loc[[0, 1], ["A", "B"]] = 1
```

File: pandas/core/indexing.py (hunk 6 above)
Comment (howsiwei) [reply pandas/core/indexing.py]: Not sure how to type key as it can be any listlike type. I was just following the type of `_get_listlike_indexer` which leaves key untyped too.

Comment (jreback) [reply pandas/core/indexing.py]: well, it would be really great to type the indexing keys

```Indexable=Iterable``` might be enough for now (you can add/import from pandas._typing)

can this be a scalar?

Comment (howsiwei) [reply pandas/core/indexing.py]: > can this be a scalar?

No.

File: pandas/core/indexing.py (hunk 4 above)
Comment (howsiwei) [reply pandas/core/indexing.py]: > then do this in the _LocIndexer (and call super); we don't want to dispatch on self.name like this (and .ix is going away shortly anyways).

Done.

Comment (howsiwei) [reply pandas/core/indexing.py]: > bigger questions, is why is this needed if you are _already_ doing this for frames above.
> 
> Exactly which test hits this?

See below.

**Comment by howsiwei** [comment]
@jreback any more changes required?

File: pandas/core/indexing.py [hunk 8]
```diff
@@ -1733,6 +1734,37 @@ def _getitem_axis(self, key, axis: int):
         self._validate_key(key, axis)
         return self._get_label(key, axis=axis)
 
+    def _ensure_listlike_indexer(self, key: Iterable, axis: int):
```
Comment (jreback) [inline pandas/core/indexing.py]: this by-definition should always be on axis=1, yes? i would remove that as an argument and/or be very explicit about this.

File: pandas/core/indexing.py [hunk 9]
```diff
@@ -1733,6 +1734,37 @@ def _getitem_axis(self, key, axis: int):
...
+        if not isinstance(self.obj._get_axis(axis), ABCMultiIndex) and all(
+            is_hashable(k) for k in key
+        ):
+            for k in key:
+                try:
+                    self.obj[k]
+                except KeyError:
+                    self.obj[k] = np.nan
+
+    def _get_setitem_indexer(self, key):
```
Comment (jreback) [inline pandas/core/indexing.py]: this *only* makes sense for a DataFrame (e.g. indexing a Series doesn't matter here).
What happens if the column indexer is a multi-index?

I would rather make these conditions much simpler and push the checks to _ensure_listlike_indexer.

The problem is the logic is currently split between this function and _ensure_listlike_indexer, which is very hard to follow why you need this.

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/indexing.py (hunk 9 above)
Comment (howsiwei) [reply pandas/core/indexing.py]: If the column indexer is a multi-index, then `_ensure_listlike_indexer` is a noop as it checks that `not isinstance(self.obj._get_axis(axis), ABCMultiIndex)`.

I split the checks between `_get_setitem_indexer` and `_ensure_listlike_indexer` because some checks are only necessary when assignment is performed using the `loc` indexer. Is it better if I pass an additional flag to `_ensure_listlike_indexer` to indicate whether to perform those checks or not?

Comment (howsiwei) [reply pandas/core/indexing.py]: If the column indexer is a multi-index, then `_ensure_listlike_indexer` is a noop as it checks that `not isinstance(self.obj._get_axis(axis), ABCMultiIndex)` before proceeding.

I split the checks between `_get_setitem_indexer` and `_ensure_listlike_indexer` because `_ensure_listlike_indexer` is also called by `DataFrame::_setitem_array` which has a different `key` format. In `DataFrame::_setitem_array` `key` is an iterable of columns. However in `_LocIndexer::_get_setitem_indexer` `key` is usually a tuple (but not always which causes the additional check) where each element of the tuple correspond to an axis (so`key[1]` is similar to `key` in `DataFrame::_setitem_array`).

File: pandas/core/indexing.py (hunk 8 above)
Comment (howsiwei) [reply pandas/core/indexing.py]: > this by-definition should always be on axis=1, yes?

Yes, `axis` is always 1.

File: pandas/core/indexing.py (hunk 9 above)
Comment (jreback) [reply pandas/core/indexing.py]: great i would prefer a single function that has all of the checks
they can be simple if/elif (rather than a compound) check

Comment (howsiwei) [reply pandas/core/indexing.py]: Like this?
```python
def _ensure_listlike_indexer(self, key, is_from_loc: bool):
    if is_from_loc:
        if (
            isinstance(key, tuple)
            and len(key) >= 2  # key is at least 2-dimensional
            and is_list_like_indexer(key[1])  # key indexes multiple columns
            and not com.is_bool_indexer(key[1])
        ):
            key = key[1]
        else:
            return

    if not isinstance(self.obj._get_axis(axis), ABCMultiIndex) and all(
        is_hashable(k) for k in key
    ):
        for k in key:
            try:
                self.obj[k]
            except KeyError:
                self.obj[k] = np.nan
```

Comment (howsiwei) [reply pandas/core/indexing.py]: Note that I do not perform the checks `is_list_like_indexer` and `not com.is_bool_indexer` when the key is coming from `DataFrame::_setitem_array` because `_setitem_array` would only be called if these condition are satisfied.

Comment (howsiwei) [reply pandas/core/indexing.py]: Updated.

File: pandas/core/indexing.py [hunk 10]
```diff
@@ -1727,6 +1728,44 @@ def _getitem_axis(self, key, axis: int):
...
+
+        Parameters
+        ----------
+        key : _LocIndexer key or list-like of column labels
+            Target labels.
+        is_indexer_key : bool
+            Whether key is a _LocIndexer key
+        """
+        column_axis = 1
+        if is_indexer_key:
```
Comment (jreback) [inline pandas/core/indexing.py]: do you *really* need is_indexer_key? this seems very odd, what exactly happens if you remove this keyword (specifically what about the below code is a problem)

File: pandas/core/indexing.py [hunk 11]
```diff
@@ -1727,6 +1728,44 @@ def _getitem_axis(self, key, axis: int):
...
+            if not (
+                isinstance(key, tuple)
```
Comment (jreback) [inline pandas/core/indexing.py]: can you put a blank line between conditions where you add a comment for each one

File: pandas/core/indexing.py [hunk 12]
```diff
@@ -1727,6 +1728,44 @@ def _getitem_axis(self, key, axis: int):
...
+            is_hashable(k) for k in key
+        ):
+            for k in key:
+                try:
+                    self.obj[k]
+                except KeyError:
+                    self.obj[k] = np.nan
+
+    def _get_setitem_indexer(self, key):
+        self._ensure_listlike_indexer(key, is_indexer_key=True)
```
Comment (jreback) [inline pandas/core/indexing.py]: ideally can you move this to the base class (so don't need to override here)

**Review by jreback — CHANGES_REQUESTED** [review]
@howsiwei thanks for sticking with it here. indexing is already *very* complicated, I am trying to ensure your change doesn't make it *even* more complicated. a few more comments.

File: pandas/core/indexing.py (hunk 12 above)
Comment (howsiwei) [reply pandas/core/indexing.py]: Where should overriding happen then? Define `_ensure_listlike_indexer` in `DataFrame` class too and override `_ensure_listlike_indexer` in `_LocIndexer`?

File: pandas/core/indexing.py (hunk 11 above)
Comment (howsiwei) [reply pandas/core/indexing.py]: Like this?
```python
if not (
    isinstance(key, tuple)
    # key is at least 2-dimensional
    and len(key) >= 2
    # key indexes multiple columns
    and is_list_like_indexer(key[column_axis])
    and not com.is_bool_indexer(key[column_axis])
):
```

File: pandas/core/indexing.py (hunk 10 above)
Comment (howsiwei) [reply pandas/core/indexing.py]: Well we need some way to differentiate whether `_ensure_listlike_indexer` is called from `DataFrame::_setitem_array` or `_LocIndexer::_get_setitem_indexer` because `key`s are in different formats depending on the caller. I have briefly explained this at https://github.com/pandas-dev/pandas/pull/29334#discussion_r359325599.

Notice that at if `is_indexer_key` is true, then I assign https://github.com/pandas-dev/pandas/blob/8ccd2b879b1a10beb7760f1f1b43fac36b41ae19/pandas/core/indexing.py#L1754 after some checks. So `is_indexer_key` is absolutely essential and the code below  the line doesn't make sense if `is_indexer_key` is removed.

File: pandas/core/indexing.py (hunk 11 above)
Comment (jreback) [reply pandas/core/indexing.py]: yes that is better

File: pandas/core/indexing.py [hunk 13]
```diff
@@ -1727,6 +1728,44 @@ def _getitem_axis(self, key, axis: int):
...
+                isinstance(key, tuple)
+                and len(key) >= 2  # key is at least 2-dimensional
+                and is_list_like_indexer(
+                    key[column_axis]
+                )  # key indexes multiple columns
+                and not com.is_bool_indexer(key[column_axis])
+            ):
+                return
+            key = key[column_axis]
+
```
Comment (jreback) [inline pandas/core/indexing.py]: add a comment explaning this

File: pandas/core/indexing.py (hunk 10 above)
Comment (jreback) [reply pandas/core/indexing.py]: i think you could remove it if you *also* check is_list_like(key).

i just find this having to separate the keywords very hard to follow.

File: pandas/core/indexing.py (hunk 12 above)
Comment (jreback) [reply pandas/core/indexing.py]: no, I mean move to _NDFrameIndexer (as opposed to _LocIndexer), then you won't have to override, you can just *always* call it in _get_setitem_indexer; that is the goal here.

we want to avoid special casing in the sub-classes of indexers, and allow _ensure_listlike_indxer to be called *always* (it just won't do anything if its conditions are not met)

close on this, thanks for working on it!

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/indexing.py (hunk 12 above)
Comment (howsiwei) [reply pandas/core/indexing.py]: Sorry, it's a typo. I wanted to say `_NDFrameIndexer` instead of `DataFrame`.

How can we ensure that conditions are only met if indexer is a `_LocIndexer`?

File: pandas/core/indexing.py (hunk 10 above)
Comment (howsiwei) [reply pandas/core/indexing.py]: `is_list_like` won't work as it returns `True` for tuple too. Checking `isinstance(key, tuple)` probably work but I'm not sure if it better reflects the intention.

File: doc/source/whatsnew/v1.0.0.rst [hunk 14]
```diff
@@ -277,6 +277,36 @@ New repr for :class:`~pandas.arrays.IntervalArray`
...
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Assignment to multiple columns of a :class:`DataFrame` when some of the columns do not exist would previously assign the values to the last column. Now, new columns would be constructed with the right values. (:issue:`13658`)
+
+.. ipython:: python
+
+   df = pd.DataFrame({'a': [0, 1, 2], 'b': [3, 4, 5]})
+   df
+
+*Previous behavior*:
```
Comment (jreback) [inline doc/source/whatsnew/v1.0.0.rst]: sp? here

File: doc/source/whatsnew/v1.0.0.rst [hunk 15]
```diff
@@ -277,6 +277,36 @@ New repr for :class:`~pandas.arrays.IntervalArray`
...
+
+   In [3]: df[['a', 'c']] = 1
+   In [4]: df
+   Out[4]:
+      a  b
+   0  1  1
+   1  1  1
+   2  1  1
+
+*New behavior*:
```
Comment (jreback) [inline doc/source/whatsnew/v1.0.0.rst]: here as well

File: pandas/core/indexing.py (hunk 10 above)
Comment (jreback) [reply pandas/core/indexing.py]: just exclude tuples

you *can* use self.name == 'loc' if you just want to handle that case.

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by jreback** [comment]
pls rebase as well

File: doc/source/whatsnew/v1.0.0.rst (hunk 14 above)
Comment (howsiwei) [reply doc/source/whatsnew/v1.0.0.rst]: What's sp?

File: pandas/core/indexing.py [hunk 16]
```diff
@@ -1963,6 +1967,47 @@ def _getitem_axis(self, key, axis: int):
...
+        Parameters
+        ----------
+        key : _LocIndexer key or list-like of column labels
+            Target labels.
+        """
+        column_axis = 1
+
+        # check if self.obj is at least 2-dimensional
+        if len(self.obj.shape) <= column_axis:
```
Comment (jreback) [inline pandas/core/indexing.py]: shouldn't this be

```
if self.obj.ndim != 2:
    return
```
?

**Review by jreback — APPROVED** [review]
lgtm! ping on green.

File: pandas/core/indexing.py (hunk 16 above)
Comment (howsiwei) [reply pandas/core/indexing.py]: i changed to
```python
if self.ndim != 2
    return
```

File: pandas/core/indexing.py (hunk 10 above)
Comment (howsiwei) [reply pandas/core/indexing.py]: @jreback Currently an error is cause by the difference between `df` and `df.loc`. Eg.
```python
df = pd.DataFrame({"a": [1, 2], "b": [3, 4]}, index=["a", "b"])
>>> df["a"]  # column "a"
a    1
b    2
Name: a, dtype: int64
>>> df.loc["a"]  # row "a"
a    1
b    3
Name: a, dtype: int64
```
How to resolve this without explicitly indicating where `key` comes from?

**Comment by TomAugspurger** [comment]
@howsiwei some CI failures now, if you have a chance to look.

**Comment by howsiwei** [comment]
@TomAugspurger I'm not sure how to fix it. See https://github.com/pandas-dev/pandas/pull/29334#discussion_r364716860

**Comment by TomAugspurger** [comment]
Thanks for the update. I'm not sure either :/

File: pandas/core/indexing.py (hunk 10 above)
Comment (howsiwei) [reply pandas/core/indexing.py]: @jreback any suggestions?

**Comment by WillAyd** [comment]
@howsiwei is this still active? Can you fix merge conflicts and try to get CI green?

Also want to move whatsnew to 1.1.0 at this point

**Comment by howsiwei** [comment]
@WillAyd I'm not sure how to proceed due to https://github.com/pandas-dev/pandas/pull/29334#discussion_r364716860. I can get the CI green with my previous approach (https://github.com/pandas-dev/pandas/blob/8ccd2b879b1a10beb7760f1f1b43fac36b41ae19/pandas/core/indexing.py#L1731) of having different checks depending on whether the DataFrame is set by `.loc[]` or `[]` but that was advised against by @jreback.

**Comment by jreback** [comment]
> @WillAyd I'm not sure how to proceed due to [#29334 (comment)](https://github.com/pandas-dev/pandas/pull/29334#discussion_r364716860). I can get the CI green with my previous approach (
> 
> https://github.com/pandas-dev/pandas/blob/8ccd2b879b1a10beb7760f1f1b43fac36b41ae19/pandas/core/indexing.py#L1731
> 
> ) of having different checks depending on whether the DataFrame is set by `.loc[]` or `[]` but that was advised against by @jreback.

you can pass axis

**Comment by simonjayhawkins** [comment]
@howsiwei can you merge master to resolve conflicts

**Comment by WillAyd** [comment]
@howsiwei can you fix up merge conflict? Otherwise should be good to merge here

**Comment by jreback** [comment]
thanks @howsiwei lgtm.

@jbrockmendel if any comments.

File: pandas/core/indexing.py [hunk 17]
```diff
@@ -611,6 +615,42 @@ def _get_setitem_indexer(self, key):
...
             raise IndexingError(key) from e
 
+    def _ensure_listlike_indexer(self, key, axis=None):
+        """
+        Ensure that a list-like of column labels are all present by adding them if
+        they do not already exist.
+
+        Parameters
+        ----------
+        key : _LocIndexer key or list-like of column labels
```
Comment (jbrockmendel) [inline pandas/core/indexing.py]: i dont think key being a _LocIndexer object makes sense

File: pandas/core/indexing.py [hunk 18]
```diff
@@ -611,6 +615,42 @@ def _get_setitem_indexer(self, key):
...
+            Target labels.
+        axis : key axis if known
```
Comment (jbrockmendel) [inline pandas/core/indexing.py]: it looks like axis isnt used here, is it necessary?

File: pandas/core/frame.py [hunk 19]
```diff
@@ -2685,6 +2685,7 @@ def _setitem_array(self, key, value):
                 for k1, k2 in zip(key, value.columns):
                     self[k1] = value[k2]
             else:
+                self.loc._ensure_listlike_indexer(key, axis=1)
```
Comment (jbrockmendel) [inline pandas/core/frame.py]: it looks like you're also calling this inside _get_setitem_indexer; do you need to also call it here?

File: pandas/core/indexing.py [hunk 20]
```diff
@@ -611,6 +615,42 @@ def _get_setitem_indexer(self, key):
...
+
+        if isinstance(key, tuple):
+            # key may be a tuple if key is a _LocIndexer key
+            # in that case, set key to the column part of key
+            key = key[column_axis]
+            axis = column_axis
+
+        if (
+            axis == column_axis
+            and not isinstance(self.obj._get_axis(column_axis), ABCMultiIndex)
```
Comment (jbrockmendel) [inline pandas/core/indexing.py]: `self.obj._get_axis(column_axis)` --> `self.obj.columns`

File: pandas/core/indexing.py [hunk 21]
```diff
@@ -611,6 +615,42 @@ def _get_setitem_indexer(self, key):
...
+            and is_list_like_indexer(key)
+            and not com.is_bool_indexer(key)
+            and all(is_hashable(k) for k in key)
```
Comment (jbrockmendel) [inline pandas/core/indexing.py]: i think the hashable check you can skip and the calls below will take care of it

File: pandas/core/indexing.py (hunk 17 above)
Comment (howsiwei) [reply pandas/core/indexing.py]: Why?  `_ensure_listlike_indexer` is called at https://github.com/pandas-dev/pandas/blob/d7184e72c9a6fcc3c565ce78dbc3e3030119a463/pandas/core/indexing.py#L586

File: pandas/core/indexing.py (hunk 18 above)
Comment (howsiwei) [reply pandas/core/indexing.py]: See https://github.com/pandas-dev/pandas/pull/29334#issuecomment-586245280

File: pandas/core/frame.py (hunk 19 above)
Comment (howsiwei) [reply pandas/core/frame.py]: `_get_setitem_indexer` is not called in this code path

File: pandas/core/indexing.py (hunk 20 above)
Comment (howsiwei) [reply pandas/core/indexing.py]: Ok, changed

File: pandas/core/indexing.py (hunk 21 above)
Comment (howsiwei) [reply pandas/core/indexing.py]: If no hashable check is performed here, some test cases fail (eg. https://github.com/pandas-dev/pandas/blob/master/pandas/tests/indexing/test_indexing.py#L112)
 because different errors are raised compared to the original code.

**Comment by jreback** [comment]
@howsiwei this looks good. can you merge master and ping on green (just to make sure)

**Comment by howsiwei** [comment]
@jreback merged master. It's green now.

File: doc/source/whatsnew/v1.1.0.rst [hunk 22]
```diff
@@ -267,6 +267,35 @@ Indexing
 - Bug in :meth:`Series.loc` and :meth:`DataFrame.loc` when indexing with an integer key on a object-dtype :class:`Index` that is not all-integers (:issue:`31905`)
 - Bug in :meth:`DataFrame.iloc.__setitem__` on a :class:`DataFrame` with duplicate columns incorrectly setting values for all matching columns (:issue:`15686`, :issue:`22036`)
 
+Assignment to multiple columns of a DataFrame when some columns do not exist
```
Comment (jreback) [inline doc/source/whatsnew/v1.1.0.rst]: move this to the section where we od api-breaking changes.

**Review by jreback — COMMENTED** [review]
@howsiwei 

all looks good, but need to move the whatsnew note. ping on green. sorry this has taken a while, we are at the finish line!

**Comment by jreback** [comment]
thanks @howsiwei very nice
thanks for sticking with it!

File: doc/source/whatsnew/v1.1.0.rst [hunk 23]
```diff
@@ -168,6 +168,37 @@ key and type of :class:`Index`.  These now consistently raise ``KeyError`` (:iss
 
 .. ---------------------------------------------------------------------------
 
+.. _whatsnew_110.api_breaking.assignment_to_multiple_columns:
+
+Assignment to multiple columns of a DataFrame when some columns do not exist
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Assignment to multiple columns of a :class:`DataFrame` when some of the columns do not exist would previously assign the values to the last column. Now, new columns would be constructed with the right values. (:issue:`13658`)
```
Comment (jbrockmendel) [inline doc/source/whatsnew/v1.1.0.rst]: "Now, new columns would" -> "Now, new columns will"

Comment (jbrockmendel) [reply doc/source/whatsnew/v1.1.0.rst]: woops, forgot this is merged, never mind
